### ISSUE TITLE
Usability issue with `opam update <package>`

### ISSUE THREAD

**Issue by gasche** [ocaml/opam#2951]
On the beginners list:

>  I also just tried `opam update utop` and got the following answer:
> > Cannot update the package utop because it is not pinned
>
> What does "pinned" mean here ?

The error message should be changed to take into account the common case where one considers `update` and `upgrade` as synonyms. I would suggest the following:

> To install the latest version of the package, you should use `opam upgrade utop` instead. The `update` command is used to update package *repositories*, or in the special case where a package is *pinned* to a development repository on your machine.

**Comment by hcarty** [comment]
I'm fairly comfortable with opam and I still make this mistake sometimes. I like @gasche's suggestion for improved wording.

**Comment by dra27** [comment]
opam could even offer to do opam upgrade in that context...

**Comment by gasche** [comment]
I think that, besides what we do when users hit the current usability bug, it would be interesting to think about how to *fix* the usability bug -- the current proposals are recovery strategies rather than fixes.

(I started writing the paragraph below with "I know this breaks compatibility so I'm not suggesting this change right now, but it's informative to lay it out still.". I'm not convinced anymore, maybe we can actually do the change right now?)

Would anything bad happen if we made `upgrade` and `update` actual synonyms? If I understand correctly, currently `update` (1) syncs repository metadata (2) syncs pin metadata and (3) runs `upgrade` (but the tool asks for confirmation so this last part can be denied). This means that the extra expressivity coming from having a separate `upgrade` is only there to upgrade stuff without synchronizing metadata first. I don't see right now why that is useful, but I think it could be relegated to an option flag on the common `update/upgrade` command: `--no-metatadata-update`?

**Comment by dra27** [comment]
It's an interesting idea - are there are other package managers which do this (off the top of my head, the `update` then `upgrade` model is very familiar - though I agree that it's trivial easy in all of them to forget which one you mean)?

There is one bonus to having the separate steps - assuming packages are cached, `opam upgrade` doesn't have to involve network access, where `opam update` always does.

I wonder if perhaps keeping the two commands, but deprecating `update` in favour of `sync`?

**Comment by dbuenzli** [comment]
> The error message should be changed to take into account the common case where one considers update and upgrade as synonyms. I would suggest the following:

I think there should be absolutely no error message here. It seems to me that the right semantics would be to simply update the OPAM repo in which the package belongs.

**Comment by dbuenzli** [comment]
(Note if you want to lump update and upgrade there's `opam update -u`)

**Comment by Drup** [comment]
My initial though was "What about scenarios without internet connection?", but actually, it doesn't really matter. It's still occasionally useful to update without upgrading, but it's pretty much pointless to upgrade without updating.

**Comment by AltGr** [comment]
Er, really ? I find this interesting and really surprising.

Indeed, I use the two independently, and for example, never use `opam update -u`. I can see many reasons why they are clearly separated in my mental model:
- first, `update` is global (for the repository part anyway), while `upgrade` is switch-specific. You can't run two repository updates in parallel. This seems a very good reason not to force an update before an upgrade to me, especially since
- `update` takes a while, esp. when you have lots of pinned packages. I often don't care about updating the pins I am not actively working on.
- their command-line arguments are completely different. I often use `upgrade` on specific packages. `update` arguments mainly allow to choose repositories, OTOH (although they also allow pinned packages, which is the confusion pointed in the OP)

In short, I only use `update` when I care about the new stuff out there. I do lots of `upgrade` when I am working, though, for my packages, pins or some dependencies. And in other switches. This is probably also related to the fact that I don't eagerly upgrade everything.

Maybe the fact that `install` and `upgrade` with arguments now automatically update the packages from the command-line if they are pinned doesn't help alleviate the confusion...

For some context, this is exactly how the venerable `apt-get` is used. Although I admit that in this case I always do `update` just before `upgrade` or `dist-upgrade`...

I also think that the distinction between the two is very important in scripts that use opam, esp. when they make use of multiple switches.

And if we force `update` before `upgrade` by default, shouldn't it be done also before `install` ?

**Comment by dbuenzli** [comment]
@AltGr what about the intial issue at hand. I suggest [this](https://github.com/ocaml/opam/issues/2951#issuecomment-305433052).

**Comment by AltGr** [comment]
Yes, saw your answer. It would make sense indeed, update the metadata and then advise `opam upgrade` as usual, gently pushing the user in the right direction rather than printing an error. It was the side-conversation about update vs. upgrade that surprised me.

**Comment by gasche** [comment]
I still suspect that this is a case where there is a real usability issue (a mismatch between user expectations and tool behavior) that is hard to see for experts familiar with the tool mental model, and that `upgrade` and `update` are too close as words to have distinct semantics. But I understand that fixing this issue may be invasive, and I think that starting with @dbuenzli's suggestion is an excellent first (mitigation) step.

**Comment by AltGr** [comment]
>  upgrade and update are too close as words to have distinct semantics

I agree on that, what surprised me is that the fact they are two distinct operations, wording apart, still seemed far from something obviously needed to many. If we were to start from scratch, probably renaming `update` to `sync` may be best. well, we could even add a `--sync` flag to package operation commands (`install`, `upgrade`...) to first do what `opam update` currently does. I'd much prefer `opam upgrade --sync` over `opam update --upgrade`.

Not sure if it would help with the ground issue, but we could also check when the last `update` was, and hint the user to use `opam sync` if it's older than a certain threshold (or e.g. to add `--sync` when running `upgrade`).

**Comment by dbuenzli** [comment]
Frankly I'm not sure `sync` would help either, as far as I'm concerned `update`, `upgrade` and `sync` could all be taken as being synonymous. 

I think the issue here is the lack of object, that is `repo` vs `pkg`.  i.e. `opam repo update` would make things clear.

**Comment by AltGr** [comment]
I quite like `opam repo update`!
But that would be equivalent to `opam update --repos`, so what should we do about updating dev/pinned packages in this case (`opam update --dev`) ? Since they are already updated, when listed on the command-line, that may be OK in general (actually, I may even like it better); yet we need a way to explicitely update them.
`opam pin update` comes to mind, and would make perfect sense, but:
- we lose the ability to update all at once
- this should only be for _pinned_ packages, not _dev_ ones (that would be packages which are not pinned, currently installed, and have been defined in the repository with a version-controlled URL. Should not happen on `opam-repository`). Maybe those could still be part of `opam repo update` (but then what about the `package` argument to update just one package) ?

`opam upgrade --sync` could do the full update before upgrading.

Finally, what should happen to `opam update` ? Display a "Warning: deprecated, use one of those instead: `opam repo update`, `opam pin update`, `opam upgrade`" ?

**Comment by kit-ty-kate** [comment]
7 years later and the error messages changed at some point (2.0.0~rc2 it seems) to:
```
$ opam update core
[ERROR] Unknown repositories or installed packages: core
```
when the package is not installed, and:
```
$ opam update ocp-indent
[WARNING] The following are not development packages (no dynamic or version controlled upstream) and can't be updated individually: ocp-indent.1.8.1
          You may want to update your repositories with just opam update or to upgrade your package with opam upgrade ocp-indent
```
when the package is installed and not pinned.

I think the original problem is thus fixed. If anyone has ideas on how to change/improve the opam UX regarding `opam update` and `opam upgrade`, feel free to open a new ticket with your ideas. Thanks!
